pkg/apis/discovery/v1beta1/conversion.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"k8s.io/api/discovery/v1beta1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/kubernetes/pkg/apis/discovery"
+)
+
+const (
+	v1beta1Topology     = "endpointslice.kubernetes.io/v1beta1-topology"
+	v1beta1ZoneTopology = "topology.kubernetes.io/zone"
+)
+
+type EndpointSliceTopology struct {
+	Strings  []string      `json:"s"`
+	Topology []map[int]int `json:"t"`
+}
+
+func Convert_v1beta1_EndpointSlice_To_discovery_EndpointSlice(in *v1beta1.EndpointSlice, out *discovery.EndpointSlice, s conversion.Scope) error {
+	// zone conversion is taken care by endpoint conversion
+	if err := autoConvert_v1beta1_EndpointSlice_To_discovery_EndpointSlice(in, out, s); err != nil {
+		return err
+	}
+
+	// Use maps initially to keep track of unique keys and counts of the values
+	keyMap := make(map[string]int)
+	valueCounts := make(map[string]int)
+	keysSize := 0
+	valuesSize := 0
+
+	// gather unique keys, values and value counts
+	for _, endpoint := range in.Endpoints {
+		for k, v := range endpoint.Topology {
+			// Don't include zone, as it will be covered by the zone field on the Endpoint
+			if k == v1beta1ZoneTopology {
+				continue
+			}
+
+			if _, ok := keyMap[k]; !ok {
+				keyMap[k] = 0
+				keysSize += len(k)
+			}
+
+			if _, ok := valueCounts[v]; !ok {
+				valueCounts[v] = 0
+				valuesSize += len(v)
+			}
+			valueCounts[v]++
+		}
+	}
+
+	approximateSize := approximateAnnotationSize(keyMap, valueCounts, len(in.Endpoints), keysSize, valuesSize)
+
+	maxSize := 256*(1<<10) - sizeOfAnnotations(out.Annotations)
+
```

Comment:
I haven't reviewed this in detail, but roundtripping API fields to json-serialized annotations in other versions has a lot of sharp edges. Some scenarios HPA hit when it took a similar path (serializing fields from an earlier API version into annotations) that you should be aware of and test for:

* Make sure the value you're marshaling/unmarshaling into the annotation is a stable API structure (HPA marshaled an internal API struct accidentally)
* Always control the annotation value for this key (by clearing it or setting it) when converting to v1
* Always drop the annotation value for this key when converting to/from v1beta1 (otherwise we could break the ability to read persisted data from etcd)

xref https://github.com/kubernetes/kubernetes/pull/89963, https://github.com/kubernetes/kubernetes/pull/89963, and https://github.com/kubernetes/kubernetes/issues/89964

Reply:
> Always control the annotation value for this key (by clearing it or setting it) when converting to v1
> Always drop the annotation value for this key when converting to/from v1beta1

Good points.  I thought there might be some trick to this, but it seems straightforward.  Ongoing discussion IRL - maybe we won't do annnotation.  More as we figure it out.